test/integration/job/job_test.go

```diff
@@ -1040,6 +1040,132 @@ func TestIndexedJob(t *testing.T) {
 	validateTerminatedPodsTrackingFinalizerMetric(t, 5)
 }
 
+func TestJobPodReplacementPolicy(t *testing.T) {
+	const initialCompletions int32 = 2
+	cases := map[string]struct {
+		featureGate          bool
+		podReplacementPolicy string
+		usePodFailurePolicy  bool
+		wantTerminating      int32
+		wantActive           int32
+	}{
+		"feature flag off, delete pods and verify no terminating status": {
+			wantTerminating: 0,
+		},
+		"feature flag true, delete pods and verify terminating status": {
+			featureGate:     true,
+			wantTerminating: 2,
+		},
+		"feature flag true, delete pods, verify terminating status and recreate upon terminating": {
+			featureGate:          true,
+			wantTerminating:      2,
+			podReplacementPolicy: string(batchv1.TerminatingOrFailed),
+		},
+		"feature flag true, delete pods, verify terminating status and recreate once failed": {
+			featureGate:          true,
+			wantTerminating:      2,
+			podReplacementPolicy: string(batchv1.Failed),
+		},
+		"feature flag false, podFailurePolicy enabled, delete pods, verify terminating status and recreate once failed": {
+			featureGate:         false,
+			usePodFailurePolicy: true,
+			wantTerminating:     2,
+		},
+	}
+	for name, tc := range cases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobPodReplacementPolicy, tc.featureGate)()
+			defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobPodFailurePolicy, tc.usePodFailurePolicy)()
+
+			closeFn, restConfig, clientSet, ns := setup(t, "pod-replacement-policy")
+			defer closeFn()
+			ctx, cancel := startJobControllerAndWaitForCaches(restConfig)
+			defer cancel()
+			resetMetrics()
+
+			// Set up initial Job in Indexed completion mode.
+			mode := batchv1.IndexedCompletion
+			jobSpec := batchv1.JobSpec{
```

Comment:
nit: I would prefer to actually make the job as part of the test cases specs. It would also allow to easily test non-indexed jobs, which I feel we need to cover.

Reply:
updated. And also added a non-indexed job test case.